assets/charts/control-plane/kubernetes/templates/etcd-endpoints.yaml

```diff
@@ -7,7 +7,7 @@ metadata:
     generatedby: bootkube
     # These labels are to make sure that the Prometheus Operator ServiceMonitor picks this
     # endpoints correctly.
-    app: prometheus-operator-kube-etcd
+    app: kube-prometheus-stack-kube-etcd
```

Comment:
This mandates the user to update the prometheus-operator immediately after cluster upgrade. And since prometheus-operator is updated the user has to update their entire infra along with components like contour, rook & metallb.

We should provide older service and endpoint along with newer one. So user can take their time to do similar changes in their infra and then decide to upgrade prometheus-operator when they need.

Reply:
@surajssd Ideally we should not advocate to the users to stop only at cluster update. A lokoctl binary should be consumed in full i.e cluster update as well as the update of components the user has installed.

What if the user is comfortable using the the component installed in previous versions and Lokomotive has moved forward 2 or more releases. A must note in our cluster upgrade process should be this "ensure all components are updated per the release notes(which should mention the components that were updated) before starting the next update process"

@invidian @iaguis Your thoughts on this ?

Reply:
>We should provide older service and endpoint along with newer one. So user can take their time to do similar changes in their infra and then decide to upgrade prometheus-operator when they need.

IMO we do not support partial upgrades and I doubt we are ready to do so. I think it's not a big deal to roll the update across all the components.

Reply:
@invidian what about the user's applications?

Reply:
> I think it's not a big deal to roll the update across all the components.

Not true for a prod cluster serving traffic.

Reply:
>@invidian what about the user's applications?

Sounds like out of scope for me. Also, we discussed in the past that this Prometheus should be treated as a "system" one.

>Not true for a prod cluster serving traffic.

I'm not saying they should be all updated at once, but until all of them are updated, I'd treat the cluster as "in maintenance window". And if something goes wrong with the update, updated components should be rolled back to previous versions to remain consistent (though we don't test that).

Reply:
I agree we shouldn't advocate being in a partially updated state and we shouldn't support that in general. However, we should make updates easy for users to apply and I think it's reasonable for users running in production to take some time to complete an update, so if we can make this easier and don't require them to update all their infra right after the control plane is update to not have a monitoring interruption we should do that.

I don't know the details in this case so feel free to ignore me if this is nonsense but maybe we can expose both endpoints here and remove the old one in the next release?

Reply:
>This mandates the user to update the prometheus-operator immediately after cluster upgrade. And since prometheus-operator is updated the user has to update their entire infra along with components like contour, rook & metallb.

- As far as I can tell upgrades should be atomic as much as possible. I actually *expect* all components which have newer versions in a release to be upgraded in a single operation. Why is it a problem to upgrade prometheus-operator, Contour, Rook and MetalLB "in one go"?
- What do you mean by "their entire infra" @surajssd? Can you name the specific things which need to be updated?
- As a user, what is the implication if the first approach is taken vs. the second approach?

Reply:
Another thought: If upgrading all components at once would have unreasonable implications from the user's perspective *and* there is nothing which can be done manually during the upgrade to mitigate these implications, another option is to release a backward-compatible update and do a "cleanup" in a later update.

Reply:
@johananl that's what I am proposing doing a backward compatible update and clean up later.

Reply:
> What do you mean by "their entire infra" @surajssd? Can you name the specific things which need to be updated?

Viz. ServiceMonitors and PrometheusRules objects across the cluster will need an update.

Reply:
This is because label `app: prometheus-operator` is now changed to `app: kube-prometheus-stack`.

This change would also escalate into any user specific applications that are being monitored.

Reply:
>This change would also escalate into any user specific applications that are being monitored and consequentially face downtime in collecting metrics.

User applications shouldn't use those labels. As far as I know, we treat shipped Prometheus instance as a system-one, integrated with components.

Reply:
> > This change would also escalate into any user specific applications that are being monitored and consequentially face downtime in collecting metrics.
> 
> User applications shouldn't use those labels. As far as I know, we treat shipped Prometheus instance as a system-one, integrated with components.
> 
> And making it backward compatible will result in duplicate targets and collected metrics :/ But still won't affect the monitoring downtime.

https://github.com/kinvolk/lokomotive/blob/master/docs/how-to-guides/monitoring-with-prometheus-operator.md#add-new-servicemonitors

https://github.com/kinvolk/lokomotive/blob/master/docs/how-to-guides/monitoring-with-prometheus-operator.md#add-custom-alerts-for-alertmanager

Reply:
https://github.com/kinvolk/lokomotive/pull/75 has a discussion about "system" Prometheus. Bummer that this is not really reflected in the docs.

Reply:
@invidian created #1204 to track this in a separate issue.